miner: exit loop when downloader Done or Failed

meowsbits opened the pull request:
Following the logic of the comment at the method,
this fixes a regression introduced at 7cf56d6f064869cb62b1673f9ee437020c595391
, which would allow external parties to DoS with
blocks, preventing mining progress.

Rel #199 


Corresponding upstream PR at https://github.com/ethereum/go-ethereum/pull/21653

meowsbits committed:
miner: exit loop when downloader Done or Failed

Following the logic of the comment at the method,
this fixes a regression introduced at 7cf56d6f064869cb62b1673f9ee437020c595391
, which would allow external parties to DoS with
blocks, preventing mining progress.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: remove ineff assign (lint)

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: update test re downloader events

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Revert "miner: remove ineff assign (lint)"

This reverts commit eaefcd34ab4862ebc936fb8a07578aa2744bc058.

meowsbits committed:
Revert "miner: exit loop when downloader Done or Failed"

This reverts commit 23abd34265aa246c38fc390bb72572ad6ae9fe3b.

meowsbits committed:
miner: add test showing imprecise TestMiner

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: fix waitForMiningState precision

This helper function would return an affirmation
on the first positive match on a desired bool.

This was imprecise; it return false positives
by not waiting initially for an 'updated' value.

This fix causes TestMiner_2 to fail, which is
expected.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: remove TestMiner_2 demonstrating broken test

This test demonstrated the imprecision of the test
helper function waitForMiningState. This function
has been fixed with 6d365c2851, and this test test
may now be removed.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: fix test regarding downloader event/mining expectations

See comment for logic.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: add test describing expectations for downloader/mining events

We expect that once the downloader emits a DoneEvent,
signaling a successful sync, that subsequent StartEvents
are not longer permitted to stop the miner.

This prevents a security vulnerability where forced syncs via
fake high blocks would stall mining operation.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: use 'canStop' state to fix downloader event handling

- Break downloader event handling into event
separating Done and Failed events. We need to
treat these cases differently since a DoneEvent
should prevent the miner from being stopped on
subsequent downloader Start events.

- Use canStop state to handle the one-off
case when a downloader first succeeds.

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: improve comment wording

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: start mining on downloader events iff not already mining

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: refactor miner update logic w/r/t downloader events

This makes mining pause/start logic regarding downloader
events more explicit. Instead of eternally handling downloader
events after the first done event, the subscription is closed
when downloader events are no longer actionable.

Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on miner/miner.go:
```diff
@@ -80,47 +80,66 @@ func New(eth Backend, config *Config, chainConfig ctypes.ChainConfigurator, mux
-		case ev := <-events.Chan():
+		case ev := <-downloaderEvents.Chan():
 			if ev == nil {
 				return
```
@meowsbits I think that if it returns at this point it will break the code, as also the `startCh` and `stopCh` will not be handled too. Maybe we can change this with a `continue`.

meowsbits replied:
Nice catch, thanks! I've added a test for this case and fixed handling the subscription events by using a dedicated go routine to avoid a busy loop. https://github.com/etclabscore/core-geth/pull/200/commits/990fc669c81b958e579e1500835fc14fe79832b9

meowsbits committed:
miner: fix handling downloader events on subcription closed

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: (lint:gosimple) use range over chan instead of for/select

Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
miner: refactor update loop to remove race condition

The go routine handling the downloader events handling
vars in parallel with the parent routine, causing a
race condition.

This change, though ugly, remove the condition while
still allowing the downloader event subscription to be
closed when the miner has no further use for it (ie DoneEvent).

iquidus reviewed (approved):
lgtm
